// __driver/post/section/finish.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache"),
	__mongodb__ = require("mongodb");

module.exports = {
	exec : function(parms){
		var fin = new Date();

		if(parms.data_sections){
			for(var v in parms.data_sections){
				var data = {
					actual_section : null
				};

				data["sections."+parms.data_sections[v].pos_section+".finished_at"] = fin;

				__db__.update({
					collection : "user_doing_survey",
					condition : {
						_id : parms.data_sections[v].id_survey_applying
					},
					set : data,
					callback : function(error, result){
						__notifications__.notifyClosingSection({
							idsurvey : parms.data_sections[v].id_survey_applying,
							idsection : parms.data_sections[v].id_section
						});
					}
				});
			}
		}else{
			__db__.find({
				collection : "user_doing_survey",
				condition : {
					_id : __mongodb__.ObjectId(parms.data.id_survey_applying)
				},
				callback : function(err, resultado){
					if(err || resultado.length != 1){

					}else{
						//encontrado
						resultado = resultado[0];
						var idSection = parms.url_data[2];
						var esta = false;
						var i = -1;
						var n = resultado.sections.length;
						while(!esta && ++i < n)
							esta = resultado.sections[i]._id == idSection;
						if(esta){
							var data = {
								actual_section : null
							};

							data["sections."+i+".finished_at"] = fin;

							__db__.update({
								collection : "user_doing_survey",
								condition : {
									_id : __mongodb__.ObjectId(parms.data.id_survey_applying)
								},
								set : data,
								callback : function(error, result){
									__notifications__.notifyClosingSection({
										idsurvey : parms.data.id_survey_applying,
										idsection : idSection
									});
									parms.response.writeHead(200, {"Content-Type": "application/json"});
									parms.response.end(JSON.stringify(result));
								}
							});
						}
					}
				}
			});
		}
	}
}
