// __driver/post/survey/closing.js
"use strict";

var mongodb = require("mongodb");
var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

function closing(parms){
	var idresource, condition, tmp;

	if(parms.url_data){
		idresource = mongodb.ObjectId(parms.url_data[0]);
		condition = {
			_id : (idresource)
		};
	}else{
		tmp = [];

		for(var v in parms.ids)
			tmp.push({_id : mongodb.ObjectId(parms.ids[v])})  ;
		condition = {
			$or : tmp
		};
	}

	var __cascade__ = new __async__([
		function(){
			__db__.update({
				collection : "survey",
				set : {
					status : "finished",
					finished_at : new Date()
				},
				condition:condition,
				callback : function(x, y){
					if(x){
						if(parms.url_data){
							parms.response.writeHead(400, {"Content-Type": "application/json"});
							parms.response.end('{}');
						}else{
							/**/
							console.log(x);
						}
					}else{
						__cascade__.next();
					}
				}
			});
		},
		function(){
			if(parms.url_data){
				__notifications__.notifyFinishedSurvey({
					idsurvey : (idresource),
					request : parms.request
				});
				parms.response.writeHead(200, {"Content-Type": "application/json"});
				parms.response.end('{}');
			}else{
				/**/
				for(var i in tmp)
					__notifications__.notifyFinishedSurvey({
						idsurvey : tmp[i]._id,
					});
			}
		}
	]).start();
}

module.exports = {
	exec: closing
}
