// __driver/post/account-recovering.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

module.exports = {
	exec: function (parms) {
		var users;
		var email = parms.data.email;
		var __cascade__ = new __async__([
			/*******************************************/
			function(){
				__db__.find({
					collection : "user",
					condition : {
						email : email
					},
					callback : function(error, u){
						users = u;
						if(users.length == 1)
							__cascade__.next();
						else{
							parms.response.writeHead(400);
							parms.response.end();
						}
					}
				});
			},
			/*******************************************/
			function(){
				var hash_recover = String(Math.random()).substr(2);
				__db__.insert({
					collection : "user_recovering",
					document : {
						iduser : users[0]._id,
						hash : hash_recover
					},
					callback : function(x, y){
						if(x){
						}else{
							var nodemailer = require('nodemailer');
							var transporter = nodemailer.createTransport('smtps://jrosendo2k%40gmail.com:n40m1j43g3r@smtp.gmail.com');
							var mailOptions = {
								from: 'admin@censo.com',
								to: email,
								subject: 'Recuperar Cuenta',
								text: 'asdf',
								html: 'http://localhost:8081/account-recovering/'+hash_recover
							};
							transporter.sendMail(mailOptions, function(error, info){
								if(error){
									//return console.log(error);
								}
								parms.response.writeHead(200, {"Content-Type": "application/json"});
								parms.response.end("{}");
							});
						}
					}
				});
			}
		]).start();
	}
}
