// __driver/post/login.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

module.exports = {
	exec: function (parms) {
		var email = parms.data.email;
		var password = parms.data.password;
		var __cascade__ = new __async__([
			//find user with that email
			function(){
				__db__.find({
					condition : {
						email : email
					},
					collection : "user",
					callback : function(error, collections){
						var passwordHash = require('password-hash');
						//if hash password true
						if(collections.length == 1 && passwordHash.verify(password, collections[0].password)){
							__cascade__.next(collections);
						}
						else{
							console.log(parms.data.email);
							parms.response.writeHead(400);
							parms.response.end();
						}
					}
				});
			},
			//restore user status to normal
			function(collections){
				__db__.update({
					condition : {
						_id : collections[0]._id
					},
					collection : "user",
					set : {
						status : "normal"
					},
					callback :	function(x, y){
						parms.request.session.logged = true;
						parms.request.session.iduser = collections[0]._id;
						parms.request.session.user_data = {
							type : collections[0].type,
							lng : "es"
						}
						__cascade__.next();
					}
				});
			},
			//save database register user session
			function(){
				__db__.insert({
					collection : "user_session",
					document : {
						iduser : parms.request.session.iduser
					},
					callback : function(x, y){
						parms.request.session.id_session_db = y._id;
						parms.response.writeHead(200, {"Content-Type": "application/json"});
						parms.response.end("{}");
					}
				});
			}
		]).start();
	}
}
