// __driver/post/master-survey-type-field.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

/*
	inserta nuevo tipo de pregunta (dato maestro)
*/

module.exports = {
	exec: function (parms) {
		var name = parms.data.name;
		var code = parms.data.code;
		var description = parms.data.description;
		var document = {}
		document["name"] = {};
		document.name[parms.request.session.user_data.lng] = name;
		document["code"] = code;
		document["description"] = {};
		document.description[parms.request.session.user_data.lng] = description;
		__db__.insert({
			collection : "master_survey_field_type",
			document : document,
			callback : function(x, y){
				if(x){
					console.log(x);
					parms.response.writeHead(400, {"Content-Type": "application/json"});
					parms.response.end("{}");
				}else{
					parms.response.writeHead(201, {"Content-Type": "application/json"});
					parms.response.end(JSON.stringify({
						item : {
							id : y._id
						}
					}));
				}
			}
		});
	}
}
