// __driver/post/register.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

module.exports = {
	exec: function (parms) {
		var email = parms.data.email.trim();
		var password = parms.data.password;
		var fullname = parms.data.fullname.trim();
		var user, hash_confirm;
		function validateEmail(email) {
			var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
			return re.test(email);
		}
		//comprobar si data es valida dentro de los parametros que se establezcan
		if(!validateEmail(email) || password.length == 0 || fullname.length == 0 || fullname.length > 512){
			parms.response.writeHead(400, {"Content-Type": "application/json"});
			parms.response.end("{}");
		}else{
			//comprobar si usuario ya existe
			var __cascade__ = new __async__([
				/*******************************************/
				function(){
					__db__.find({
						collection : "user",
						condition : {
							email : email
						},
						callback : function(e, ct){
							if(ct.length == 0){
								var passwordHash = require('password-hash');
								password = passwordHash.generate(password);
								__cascade__.next();
							}else{
								parms.response.writeHead(409, {"Content-Type": "application/json"});
								parms.response.end("{}");
							}
						}
					});
				},
				/*******************************************/
				//insercion de usuario en la respectiva coleccion, con status "por confirmar"
				function(){
					__db__.insert({
						collection : "user",
						document : {
							email:email,
							password:password,
							fullname:fullname,
							status : "confirm",
							type : "admin"
						},
						callback : function(error, u){
							user = u;
							if(error){
							}else{
								hash_confirm = String(Math.random()).substr(2);
								__cascade__.next();
							}
						}
					});
				},
				/*******************************************/
				//cargar registro de usuario por confirmar en la respectiva tablas
				function(){
					__db__.insert({
						collection : "user_confirm",
						document : {
							iduser : user._id,
							hash : hash_confirm
						},
						callback : function(err, confirm){
							//enviar email de confirmacion
							var nodemailer = require('nodemailer');
							var transporter = nodemailer.createTransport('smtps://jrosendo2k%40gmail.com:8088yp4g3@smtp.gmail.com');
							var mailOptions = {
								from: 'admin@censo.com',
								to: email,
								subject: 'Confirmar Registro',
								text: 'asdf',
								html: 'http://localhost:8081/register-confirmation/'+hash_confirm
							};
							transporter.sendMail(mailOptions, function(error, info){
								if(error){
								}
								parms.response.writeHead(201, {"Content-Type": "application/json"});
								parms.response.end("{}");
							});
						}
					});
				}
			]).start();
		}
	}
};
