// __driver/notifications/index.js
"use strict";

var io;

function main(server){
	io = require('socket.io')(server);
	io.on('connection', function(socket){
		//console.log("somebody connected...");
		//console.log(socket);
	});
}

//io.sockets.emit('notification_survey_activated', Math.random());

function notifyStartedSurvey(parms){
	//cargar notificaciones en base de datos (analisis incompleto, por ahora)
	//enviar notificaciones por sockets relacionadas a base de datos
	//enviar otros tipos de pushing
	io.sockets.emit('notification_survey_'+parms.idsurvey+'_activated', {
		//empty data
	});
}

function notifyFinishedSurvey(parms){
	//cargar notificaciones en base de datos (analisis incompleto, por ahora)
	//enviar notificaciones por sockets relacionadas a base de datos
	//enviar otros tipos de pushing
	io.sockets.emit('notification_survey_'+parms.idsurvey+'_finished', {
		//empty data
	});
}

function notifyRemovedSurvey(parms){
	//cargar notificaciones en base de datos (analisis incompleto, por ahora)
	//enviar notificaciones por sockets relacionadas a base de datos
	//enviar otros tipos de pushing
	io.sockets.emit('notification_survey_'+parms.idsurvey+'_removed', {
		//empty data
	});
}

function notifyClosingSection(parms){
	io.sockets.emit('notification_survey_'+parms.idsurvey+'_section_'+parms.idsection+'_closing', {
		//empty data
	});
}

module.exports = {
	start : main,
	notifyStartedSurvey : notifyStartedSurvey,
	notifyFinishedSurvey : notifyFinishedSurvey,
	notifyRemovedSurvey : notifyRemovedSurvey,
	notifyClosingSection : notifyClosingSection
}
