// __driver/get/account-recovering.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

/*
	controlador encargado procesar la recuperacion de cuenta (cuando el usuario accede al link de recuperacion
	que se le envia como email a su bandeja de entrada)
*/

module.exports = {
  exec: function (parms) {
  	var collections;
  	var __cascade__ = new __async__([
		/*************************/
		//verifica si el hash dado en la url existe
		function(){
			__db__.find({
		  		condition : {
		  			hash : parms.data[0]
		  		},
		  		collection : "user_recovering",
		  		callback : function(e, c){
		  			collections = c;
					if(c.length == 1){
						__cascade__.next(c[0].iduser);
					}else{
						parms.response.writeHead(404);
						parms.response.end();
					}
				}
			});
		},
		/*************************/
		//en caso de que exista actualizar normalizar el status del usuario
		function(iduser){
			__db__.update({
				collection : "user",
				condition : {
					_id : iduser
				},
				set : {
					status : "normal"
				},
				callback :  function(err, results){
					if(err){
					}else{
						__cascade__.next(iduser);
					}
				}
			});
		},
		/*************************/
		//eliminar registro del usuario con la solicitud de recuperacion de cuenta
		function(iduser){
			__db__.delete({
				collection : "user_recovering",
				condition : {
					_id : collections[0]._id
				},
				callback : function(err, results){
					//inicio automatico de sesion
					parms.request.session.logged = true;
					parms.request.session.iduser = iduser;
					parms.request.session.user_data = {
						type : collections[0].type,
						lng : "es"
					};
					__cascade__.next();
				}
			});
		},
		/*************************/
		//registrar inicio de sesion en la base de datos
		function(){
			__db__.insert({
				collection : "user_session",
				document : {
					iduser : parms.request.session.iduser
				},
				callback : function(x, y){
					parms.request.session.id_session_db = y._id;
					parms.response.writeHead(301, {
						Location: "/"}
					);
					parms.response.end();
				}
			});
		}
	]).start();
  }
}
