// __driver/get/account-disabling.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

/*
	desactiva la cuenta de un usuario
*/

module.exports = {
	exec: function (parms) {
		var iduser = parms.request.session.iduser;
		var mongodb = require('mongodb');
	  	var __cascade__ = new __async__([
	  		/*********************************/
	  		//actualiza el status del usuario (previo deberia validarse que haya una sesion abierta)
	  		function(){
				__db__.update({
					collection : "user",
					condition : {
						_id : mongodb.ObjectID(iduser)
					},
					set : {
						status : "disabled"
					},
					callback : function(x, y){
						if(x){
						}else{
							__cascade__.next();
						}
					}
				});
	  		},
	  		/*********************************/
	  		//cierre de sesion (en node y en base de datos)
	  		function(){
				__db__.delete({
					collection : "user_session",
					condition : {
						_id : mongodb.ObjectID(parms.request.session.id_session_db)
					},
					callback : function(error, result){
						parms.request.session.destroy();
						parms.response.writeHead(301, {
							Location: "/"}
						);
						parms.response.end();
					}
				});
	  		}
	  	]).start();
	}
}
