// __driver/get/admin.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

module.exports = {
	exec : function(request, response){
		if(request.session.logged){
			var parms = {
				type_user : request.session.user_data.type,
				nameSection : "admin"
			};
			var async = new __async__([
				function(){
					__db__.find({
						collection : "master_survey_field_type",
						callback : function(error, types){
							var ts = Array();
							for(v in types){
								ts.push({
									code :  types[v]["code"],
									name : types[v]["name"]["es"],
									description : types[v]["description"]["es"],
									id : types[v]["_id"]
								});
							}
							parms["field_types"] = ts;
							async.next();
						}
					});
				},
				function(){
					__db__.find({
						collection : "survey",
						condition : {
							iduser : request.session.iduser
						},callback:function(error, surveys){
							for(v in surveys){
								surveys[v]["stringify_tags"] = JSON.stringify(surveys[v]["tags"]);
							}
							parms["surveys"] = surveys;
							response.render("index", parms);
						}
					});
				}
			]).start();  
		}else{
			response.writeHeader(301, {
				Location: "/"
			});
			response.end();
		}
	}
}
