// __driver/get/register-confirmation.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

//controlador programado para procesar la confirmacion de registro por parte de un usuario

module.exports = {
  exec: function (parms) {
  	var iduser;
  	var collections;
	var __cascade__ = new __async__([
		/*******************************************/
		//verificar si existe en la coleccion respectiva un documento con el hash proveido
		function(){
		  	__db__.find({
		  		condition : {
		  			hash : parms.data[0]
		  		},
		  		collection : "user_confirm",
		  		callback : function(error, c){
		  			collections = c;
					if(collections.length == 1){
						iduser = collections[0].iduser;
						__cascade__.next();
					}else{
						parms.response.writeHead(400);
						parms.response.end();
					}
				}
			});
		},
		/*******************************************/
		//en caso de que exista el documento con el hash, marcar al usuario asociado con status de "normal"
		function(){
			__db__.update({
				collection : "user",
				condition : {
					_id : iduser
				},
				set : {
					status : "normal"
				},
				callback :  function(err, results){
					if(err){
						parms.response.writeHead(400);
						parms.response.end();
					}else{
						__cascade__.next();
					}
				}
			});
		},
		/*******************************************/
		//crear sesion, iniciarla
		function(){
			__db__.delete({
				collection : "user_confirm",
				condition : {
					_id : collections[0]._id
				},
				callback : function(err, results){
					parms.request.session.logged = true;
					parms.request.session.iduser = iduser;
					parms.request.session.user_data = {
						type : collections[0].type,
						lng : "es"
					};
					__cascade__.next();
				}
			});
		},
		/*******************************************/
		//insertar registro de inicio de sesion en base de datos
		function(){
			__db__.insert({
				collection : "user_session",
				document : {
					iduser : parms.request.session.iduser
				},
				callback : function(x, y){
					parms.request.session.id_session_db = y._id;
					parms.response.writeHead(301, {
						Location: "/"}
					);
					parms.response.end();
				}
			});
		}
	]).start();
  }
}
