// __driver/delete/survey.js
"use strict";

var __db__ = require('__driver/db'),
	__async__ = require('__driver/myasync').async,
	__notifications__ = require("__driver/notifications"),
	__cache__ = require("__driver/cache");

module.exports = {
	exec: function (parms) {
		var mongodb = require("mongodb");
		var idresource = parms.url_data[0];
		var __cascade__ = new __async__([
			function(){
		  		/*********************************/
		  		//eliminar cuestionario de la base de datos
		  		//por ahora solo esta implementado la eliminacion manual por parte del usuario
		  		//quizas despues la eliminacion automatica, por parte del sistema o automatizacion
		  		//establecida por el usuario, lo cual cambiaria, agregaria y/o eliminaria algunos parametros
				__db__.delete({
					collection : "survey",
					condition:{
						_id : mongodb.ObjectId(idresource)
					},
					callback : function(x, y){
						if(x){
							parms.response.writeHead(400, {"Content-Type": "application/json"});
							parms.response.end('{}');
						}else{
							__cascade__.next();
						}
					}
				});	
			},
			function(){
				//eliminar cuestionario de la cache en caso de que se encuentre ahi
				//para evitar cargas innecesarias en el sistema
				__cache__.del({
					category : {
						name : "survey",
						category : {
							name : "not-started"
						}
					},
					key : idresource
				});
				__cache__.del({
					category : {
						name : "survey",
						category : {
							name : "ongoing"
						}
					},
					key : idresource
				});
  				//enviar notificaciones de cuestionario removido
  				//esto es util cuando alguien esta navegando en una encuesta pero el usuario dueño la elimina
				__notifications__.notifyRemovedSurvey({
					idsurvey : mongodb.ObjectId(idresource),
					request : parms.request
				});
				parms.response.writeHead(200, {"Content-Type": "application/json"});
				parms.response.end('{}');
			}
		]).start();
	}
}
